client/js/widgets/dice.js

```diff
@@ -0,0 +1,419 @@
+      classes: 'shaking',
+      clickable: true,
+      movable: false,
+      layer: 1,
+
+      borderRadius: '16%',
+      pipColor: 'black',
+      backgroundColor: 'white',
+      borderColor: null,
+
+      faces: [ 1, 2, 3, 4, 5, 6 ],
+      activeFace: 0,
+      rollCount: 0,
+
+      rollTime: 800,
+      swapTime: 500,
+
+      image: null,
+      text: null,
+      pips: null,
+      svgReplaces: null,
+      faceCSS: null,
+
+      pipSymbols: null,
+      shape3d: false
+    });
+  }
+
+  activeFace() {
+    return mod(Math.round(this.get('activeFace')), this.faceElements.length) || 0;
+  }
+
+  animate() {
+    if(!this.facesElement || !this.faceElements.length)
+      return;
+
+    const faceElements = this.faceElements;
+
+    const hash = this.rollHash ? this.rollHash : 0;
+    const fc = faceElements.length;
+    const af = this.activeFace();
+    const f0 = this.previousActiveFace;
+    let f1 = (f0 + (hash) % (fc - 1) + 1) % fc || 0;
+    if(f1 == f0)
+      f1 = (f1 + 1) % fc;
+    let f2 = (f1 + (hash >>> 14) % (fc - (f1 == af ? 1 : 2)) + 1) % fc || (f1 + 1) % fc;
+    if(f2 == af)
+      f2 = (f2 + 1) % fc;
+
+    this.facesElement.className = `diceFaces shake${hash >>> 29}`;
+
+    for(const face of faceElements)
+      face.classList.remove('animate1active', 'animate2active');
+    faceElements[f1].classList.add('animate1active');
+    faceElements[f2].classList.add('animate2active');
+  }
+
+  animateProperties() {
+    const rules = super.animateProperties();
+    const swap = this.get('swapTime');
+    const roll = this.get('rollTime');
+    rules.unshift(
+      { property: 'activeFace', duration: typeof swap == 'number'? swap : 1000},
+      { property: 'rollCount', duration: typeof roll == 'number'? roll + 50 : 1000},
+      { property: 'rollCount', className: 'animateBegin', duration: 50 }
+    );
+    return rules;
+  }
+
+  applyDeltaToDOM(delta) {
+    const diceDelta = {};
+    if(delta.activeFace !== undefined || delta.faces !== undefined) {
+      const priorFace = this.previousActiveFaceProps;
+      if(typeof priorFace == 'object' && priorFace != null) {
+        for(const prop in priorFace) {
+          if(Object.prototype.hasOwnProperty.call(priorFace, prop))
+            diceDelta[prop] = null;
+        }
+      } else {
+        diceDelta['value'] = null;
+      }
+    }
+    Object.assign(diceDelta,delta);
+    if(delta.activeFace !== undefined || delta.faces !== undefined) {
+      const face = this.faces()[this.get('activeFace')];
+      if(typeof face == 'object' && face != null)
+        Object.assign(diceDelta,face);
+      else
+        diceDelta['value'] = face;
+      this.previousActiveFaceProps = face;
+    }
+    delta = diceDelta;
+
+    super.applyDeltaToDOM(delta);
+
+    let childNodesWereRecreated = false;
+    if([ 'faces', 'shape3d', 'pipSymbols', 'faceCSS', 'image', 'text', 'pips', 'svgReplaces' ].reduce((result,p)=>(result||delta[p]!==undefined),false)) {
```

Comment:
If these property changes are from a face change (or if all faces override the widget property anyway), the resulting recreated nodes may be identical to the existing nodes, but current code does maintain consistency across clients.